python/kserve/kserve/protocol/rest/openai/openai_proxy_model.py

```diff
@@ -0,0 +1,230 @@
+from typing import AsyncIterator, Optional, Union
+import httpx
+from functools import partial
+import orjson
+
+
+from .openai_model import (
+    OpenAIModel,
+    AsyncMappingIterator,
+    CompletionRequest,
+    ChatCompletionRequest,
+)
+from .types import (
+    ChatCompletion,
+    ChatCompletionChunk,
+    Completion,
+)
+from .errors import OpenAIError, create_error_response
+
+
+COMPLETIONS_ENDPOINT = "/v1/completions"
+CHAT_COMPLETIONS_ENDPOINT = "/v1/chat/completions"
+
+
+class OpenAIProxyModel(OpenAIModel):
+    """
+    An implementation of OpenAIModel that proxies requests to a backend server exposing Open AI endpoints.
+
+    Users can extend this class and override any of the following methods to hook into the request/response cycle:
+        - preprocess_completion_request
+        - postprocess_completion
+        - postprocess_completion_chunk
+        - preprocess_chat_completion_request
+        - postprocess_chat_completion
+        - postprocess_chat_completion_chunk
+
+    Each method takes a single parameter: the object currently being processed. Request objects may be mutated to modify
+    the request sent to the downstream server and response objects (completions/completion chunks) may be mutated to
+    modify the response returned to the upstream caller.
+
+    Args:
+        predictor_url (str):
+            The url of the model server to send requests to. Should be fully qualified with scheme, host, and port.
+            e.g. `http://my-backend:9000`
+        http_client (httpx.AsyncClient|None):
+            An optional instance of httpx.AsyncClient to use for sending requests to the upstream server.
+    """
+
+    predictor_url: str
+    skip_upstream_validation: bool
+    _http_client: httpx.AsyncClient
+    _completions_endpoint: str
+    _chat_completions_endpoint: str
+
+    def __init__(
+        self,
+        name: str,
+        predictor_url: str,
+        http_client: Optional[httpx.AsyncClient] = None,
+        skip_upstream_validation: bool = False,
+    ):
+        super().__init__(name)
+        self.predictor_url = predictor_url
+        self._http_client = (
+            http_client
+            if http_client is not None
+            else httpx.AsyncClient(timeout=httpx.Timeout(read=30.0))
```

Comment:
Fyi. I got this error when exercising this branch.
```
  raise ValueError(
ValueError: httpx.Timeout must either include a default, or set all four parameters explicitly.
```

Reply:
Ah, good catch. I think this isn't tested in the unit tests because we pass in a mock client and it isn't tested in my example because we explicitly set the client there. Will fix 👍